src/search/containers/CurrentSearchPage.js

```diff
@@ -7,19 +8,33 @@ export function mapStateToProps(state) {
   return state.search;
 }
 
-export function mapDispatchToProps(dispatch) {
-  return {
-    handleSearch: (query, page = 1) => {
-      dispatch(searchStart(query, page));
-      return search({ page, query })
-        .then((response) => dispatch(searchLoad({ page, query, ...response })));
-    },
-  };
+function performSearch({dispatch, page, query}) {
+  dispatch(searchStart(query, page));
+  return search({ page, query })
+    .then((response) => dispatch(searchLoad({ page, query, ...response })));
 }
 
-const CurrentSearchPage = connect(
-  mapStateToProps,
-  mapDispatchToProps,
-)(SearchPage);
+export function isLoaded({page, query, state}) {
+  return state.query === query && state.page === page && !state.loading;
+}
+
+export function parsePage(page) {
+  const parsed = parseInt(page, 10);
+  return Number.isNaN(parsed) || parsed < 1 ? 1 : parsed;
+}
+
+export function loadSearchResultsIfNeeded({store: {dispatch, getState}, location}) {
+  const query = location.query.q;
+  const page = parsePage(location.query.page);
+  if (!isLoaded({state: getState().search, query, page})) {
+    return performSearch({dispatch, page, query});
+  }
+  return true;
```

Comment:
It appears to work. I would assume that redux-async-connect is doing something like `Promise.all(asyncConnectPromised).then(continueToRedux)` which will wrap values in promises. 

The reason I return immediately is so that i know nothing async is happening when testing this.